Plots/Polygons/NCL_polyg_2.py

```diff
@@ -0,0 +1,140 @@
+  - Drawing a custom labelbar on a map
+  - Creating a red-yellow-blue color map
+
+See following URLs to see the reproduced NCL plot & script:
+    - Original NCL script: https://www.ncl.ucar.edu/Applications/Scripts/polyg_2.ncl
+    - Original NCL plot: https://www.ncl.ucar.edu/Applications/Images/polyg_2_lg.png
+"""
+
+###############################################################################
+# Import packages:
+# ----------------
+import xarray as xr
+import cartopy.crs as ccrs
+import matplotlib.pyplot as plt
+import matplotlib.colors as colors
+import matplotlib.cm as cm
+from geocat.viz import util as gvutil
+
+import geocat.datafiles as gdf
+import cartopy.io.shapereader as shpreader
+import shapely.geometry as sgeom
+from mpl_toolkits.axes_grid1.inset_locator import inset_axes
+
+###############################################################################
+# Helper function to determine which color to fill the divisions based on precipitation data
+
+
+def findDivColor(colormap, pdata):
+    for x in colormap:
+        if pdata >= x[0]:
+            continue
+        else:
+            return x[1]
+
+
+###############################################################################
+# Plot map and colorbar
+
+# Create plot figure
+fig = plt.figure(figsize=(10, 10))
+
+# Add a subplot for lambert conformal map
+ax = fig.add_subplot(111, projection=ccrs.LambertConformal(), frameon=False, xbound=0.0, ybound=0.0)
+
+# Set latitude and longitude extent of map
+ax.set_extent([-119, -74, 18, 50], ccrs.Geodetic())
+
+# Set shape name of map (which depicts the United States)
+shapename = 'admin_1_states_provinces_lakes_shp'
+states_shp = shpreader.natural_earth(resolution='110m', category='cultural', name=shapename)
+
+# Set title and title fontsize of plot using gvutil function instead of matplotlib function call
+gvutil.set_titles_and_labels(ax, maintitle="\nAverage Annual Precipiation \n Computed for the period 1899-1999 \n NCDC climate division data \n \n",
+                             maintitlefontsize=18)
+
+# Add outlines of each state within the United States
+for state in shpreader.Reader(states_shp).geometries():
+
+    facecolor = 'white'
+    edgecolor = 'black'
+
+    ax.add_geometries([state], ccrs.PlateCarree(), facecolor=facecolor, edgecolor=edgecolor)
+
+# Open climate division datafile and add to xarray
+ds = xr.open_dataset(gdf.get("netcdf_files/climdiv_prcp_1899-1999.nc"), decode_times=False)
+
+# For each variable (climate division) in data set, create outline on map and fill with random color
+for varname, da in ds.data_vars.items():
+
+    try:
+        # Get precipitation data for each climate division
+        first = da
+        precipitationdata = sum(first.values)/100
+
+        # Get borders of each climate division
+        lat = first.lat
+        lon = first.lon
+
+        # Set color map with bounds
+        colormap = [(5, 'mediumpurple'), (10, 'mediumblue'), (15, 'royalblue'),
```

Comment:
Now we have the list of colors duplicated (here and after the for loop when you generate your ListedColormap), and we have the mapping interval  (5, 10, 15, etc.) duplicated here and when you create the 'bounds' loop. How can you restructure this code so that you can still have a help function for mapping a color to an index, but avoid the duplication of values/colors? The problem with duplication like this is that it leads to less maintainable code; someone, for example, will change the 10 in colormap, but forget to change it in bounds.

Reply:
Fixed it! Now there is only one colormap dictionary defined, and the other two temporary arrays draw their color/bound values from it.

Reply:
Hmm. Moving in the right direction, but...

Now we have two colormaps, and two sets of bounds, where one of each would easily do. Rather than creating a custom, dictionary colormap that will be unfamiliar to readers we can make do with the single matplotlib.colors class to store only our colors, and a single array to store our bounds. All we need to do is generate the matplotlib.colors instance from the list of colors, create a bounds array, and take advantage of the fact that ListedColormaps store their colors - the color names that you've provided - in the .colors attribute. Hence, you can access the color names in the 'cmap' variable with:

`cmap.colors[i]`

where 'i' is the color index. Then you can make a simple change to findDivColor so that it operates on the 'bounds' array and returns an index that you can then use to retrieve the desired color. Make sense? The point of all this is to avoid having to create separate copies of the colormap and bounds information in different representations when one will easily do

Reply:
Looks good!